### ISSUE TITLE
Differing Speeds as Compared To The Paper After Distillation

### ISSUE THREAD

**Issue by mahanswaray** [microsoft/Distilled-Sentence-Embedding#2]
I attempted to reproduce your numbers for the NLI task using the NLI corpus but I'm not getting the same computation speed up as outlined in the paper

SENTENCE EMBEDDING EXTRATION ON CPU:
BERT-LARGE-WHOLE-WORD-MASKING took 74s to encode 1000 sentences one by one (batchSize = 1)
DSE Trained on logits from finetuned model of BERT-LARGE-WHOLE-WORD-MASKING took 67s  to encode 1000 sentences one by one (batchSize = 1)

This speed up is quite low and compared to the paper it is nowhere even close

MORE INFO

I followed the example python commands to the dot.

a) Finetuned a large bert model using the script for MNLI corpus

b) Extracted Logits using the script for MNLI corpus

c) Trained a DSE model using the logits 
    i) Hardware specifications : A single 2080TI 11GB CUDA RAM, VM with 12 vCPU cores and 32 GB system RAM
    ii) Hyperparameters - All same as default except batchSize is 4 (Had to reduce because 11GB CUDA ram was a constraint )
          - As a reference at batchsize 4 9GB of ram was being consumed just by the models.
    iii) Took a total of 80 hrs to train on 8 epochs 

d) Once the model was trained i used the `load_dse_checkpoint_example.py` to load the dse model, tokenize sentence pairs into one input_ids sequence which was fed to the dse model
i ran this same sentence embedding extraction a 1000 times in a loop in 67 s
the same 1000 runs looping around a bert large model for extracting sentence embedding took 74s 

Can you please reach out and explain if i'm doing something wrong or what could be wrong?
because sentence pair sentence embedding extraction in the paper was sped BY A FACTOR OF 934

THIS IS NOT WHAT I HAVE BEEN ABLE TO REPRODUCE. :((

**Comment by noamrazin** [comment]
Hi,

If I understand correctly, you are attempting to create sentence embeddings for 1000 sentences 1 by 1. Indeed, DSE does not give any major speed up for this scenario, and this is not what is claimed in the paper. Notice that the model size and computations are essentially the same for an original  BERT-Large model and DSE when encoding single sentences, therefore this is to be expected.
As described in the paper, the speed-up is observed in the following scenarios, common in large-scale ranking and retrieval tasks:
1. **Computing similarities for all pairs of sentences**: Displayed in Table 3 in the paper. Since DSE decouples the computation for each sentence, this is done by first computing the 1000 sentence embeddings, followed by computing 1M similarity scores using a low-cost similarity function (e.g. a shallow FC network, cosine-similarity). Notice that in this case, to compute the similarity score with the original BERT-Large model you must pass all 1M pairs through the whole network due to cross-attention operations.

2. **Computing similarities for a single query sentence and an existing large database of sentences:** Displayed in Table 4 in the paper. In this case, we assume that sentence embeddings for the existing database were pre-computed, which is possible only when the encoding process for a query and candidate sentence is decoupled (as in DSE). Thus, DSE requires only to encode the new query sentence, followed by computing 100K similarity scores using a low-cost similarity function (the query sentence is scored against all sentences in the database). Naively, BERT does not allow such a decoupling, and requires to run all 100K possible query-candidate pairs through the whole network (i.e. pre-computing embeddings for existing sentences in the database does not offer any speed-up in general).

I hope this is clear, and you will be able to reproduce the results.

**Comment by mahanswaray** [comment]
So in a case where 1m pairs of DIFFERENT sentences want to be tested that DSE and BERT-LARGE would take the same amount of time for sentence embedding extraction?

**Comment by mahanswaray** [comment]
Also not sure if i agree with you that all 1m pairs must be sent through bert model
since you could extract 1000 sentences' cls embedding and do pairwise similarity there too?

why should all 1m pairs be sent to the bert model?

the comparison itself is quite misleading and you should mention for any other task apart from sentence similarity this speed up is not seen, since you yourself said that the single extraction of one sentence's embedding is not much faster using DSE vs BERT large

**Comment by mahanswaray** [comment]
just to conclude 
one cannot expect DSE to give universal sentence embeddings to be downstreamed on any task any faster than BERT-LARGE
unless it's cross-attention similarity task?

**Comment by noamrazin** [comment]
1. In case of 1M different pairs of sentences, where all sentences are unique (2M sentences total), then DSE does not offer a significant speed-up over the standard BERT-Large. As stated in the paper, the speed up is mainly in the 2 common scenarios mentioned above.

2. Yes, you can use the standard BERT model and extract sentence embeddings from it by plugging in 1 sentence at a time (we compare to such a method in the downstream tasks section). There are multiple ways to do so --- e.g. taking the CLS representation, or some pooling operation over the token representations. However, BERT itself is pre-trained only with pairs of sentences, and we've observed this to be sub-optimal. We offer one way of alleviating this issue by using knowledge distillation to train BERT in a manner that better suits the creation of sentence embeddings for single sentences.

3. Yes, the purpose of DSE is to allow significant speed-up with a relatively small degradation in performance for tasks that require scoring pairs of sentences, where some sentences appear multiple times. Thus, simply for creating sentence representations it does not offer a significant speed-up. Yet, as shown in the paper, the quality of embeddings (according to several benchmarks) still improves over vanilla BERT.
